File: solution/1608. Special Array With X Elements Greater Than or Equal X/1608_03.cpp
class Solution {
public:
    int specialArray(vector<int>& nums) {
        const int n = nums.size();
        
        int l = 0, r = n, m;
        while (l <= r) {
            m = l + (r - l) / 2;
            int cnt = 0;
            for (int x : nums)
                cnt += x >= m;
            
            if (cnt == m)
                return m;
            else if (cnt < m)
                r = m - 1;
            else if (cnt > m)
                l = m + 1;
        }
        return -1;
    }
};

File: solution/1608. Special Array With X Elements Greater Than or Equal X/1608_03.py
class Solution:
    def specialArray(self, nums: List[int]) -> int:
        l, r = 0, len(nums)
        while l <= r:
            m = l + (r - l) // 2
            cnt = sum(x >= m for x in nums)
            if cnt == m:
                return cnt
            elif cnt > m:
                l = m + 1
            elif cnt < m:
                r = m - 1
        return -1
